lib/codegen/src/nan_canonicalization.rs

```diff
@@ -0,0 +1,120 @@
+//! A NaN-canonicalizing rewriting pass. For instructions that can potentially
+//! result in a nondeterministic NaN value, insert operations to check for NaN,
+//! and replace the result with a deterministic canonical NaN value if the
+//! result of an instruction was in fact a NaN.
+
+use cursor::{Cursor, FuncCursor};
+use ir::{DataFlowGraph, Function, Inst, InstBuilder, InstructionData, Opcode, Value};
+use ir::condcodes::FloatCC;
+use ir::immediates::{Ieee32, Ieee64};
+use ir::types;
+use ir::types::Type;
+use timing;
+
+// Canonical 32-bit and 64-bit NaN values.
+static CANON_32BIT_NAN: u32 = 0b01111111100000000000000000000001;
+static CANON_64BIT_NAN: u64 =
+    0b0111111111110000000000000000000000000000000000000000000000000001;
+
+/// Performs the NaN-canonicalization pass by identifying floating-point
+/// arithmetic operations, and adding instructions to replace the result
+/// with a canonical NaN value if the result of the operation was NaN.
+pub fn do_nan_canonicalization(func: &mut Function) {
+    let _tt = timing::canonicalize_nans();
+    let mut pos = FuncCursor::new(func);
+    while let Some(_ebb) = pos.next_ebb() {
+        while let Some(inst) = pos.next_inst() {
+            if is_fp_arith(&mut pos, inst) {
+                add_nan_canon_instrs(&mut pos, inst);
+            }
+        }
+    }
+}
+
+/// Returns true/false based on whether the instruction is a floating-point
+/// arithmetic operation. This ignores operations like `fneg`, `fabs`, or
+/// `fcopysign` that only operate on the sign bit of a floating point value.
+fn is_fp_arith(pos: &mut FuncCursor, inst: Inst) -> bool {
+    // FIXUP: Should `ceil`, `floor`, `trunc`, `nearest`, and
+    // immediate constants be considered as well?
```

Comment:
Yes for `ceil`, `floor`, `trunc`, and `nearest`, and no for immediate constants.

Reply:
Awesome, can do :)